sys/parsoid.js

```diff
@@ -519,47 +544,63 @@ class ParsoidService {
-                        body: {
-                            type: 'precondition_failed',
-                            detail: 'The precondition failed'
-                        }
-                    });
-                }
-                return generateContent(res);
-            }, generateContent);
+        const disabledStorage = this._isStorageDisabled(req.params.domain);
+
+        if (!disabledStorage) {
+            contentReq = this._getContentWithFallback(
+                hyper, rp.domain, rp.title, rp.revision, rp.tid
+            );
+
+            if (mwUtil.isNoCacheRequest(req)) {
+                // Check content generation either way
+                contentReq = contentReq.then((res) => {
+                    if (isModifiedSince(req, res)) { // Already up to date, nothing to do.
+                        throw new HTTPError({
+                            status: 412,
+                            body: {
+                                type: 'precondition_failed',
+                                detail: 'The precondition failed'
+                            }
+                        });
+                    }
+                    return generateContent(res);
+                }, generateContent);
+            } else {
+                // Only (possibly) generate content if there was an error
+                contentReq = contentReq.catch(generateContent);
+            }
         } else {
-            // Only (possibly) generate content if there was an error
-            contentReq = contentReq.catch(generateContent);
+            contentReq = this._getPageBundleFromParsoid(hyper, req);
         }
+
         return contentReq
             .then((res) => {
                 res.headers = res.headers || {};
                 if (!res.headers.etag) {
                     res.headers.etag = res.body.html.headers && res.body.html.headers.etag;
                 }
                 if (!res.headers.etag || /^null$/.test(res.headers.etag)) {
-                    // if there is no ETag, we *could* create one here, but this
-                    // would mean at least cache pollution, and would hide the
-                    // fact that we have incomplete data in storage, so error out
-                    hyper.logger.log('error/parsoid/response_etag_missing', {
-                        msg: 'Detected a null etag in the response!'
-                    });
-                    throw new HTTPError({
-                        status: 500,
-                        body: {
-                            title: 'no_etag',
-                            description: 'No ETag has been provided in the response'
-                        }
-                    });
+                    if (disabledStorage) {
+                        // Generate an ETag, for consistency
+                        const uuid = uuidv1();
+                        res.headers.etag = mwUtil.makeETag(res.headers['content-revision-id'] || '0', uuid);
```

Comment:
I think here we are missing some DOM manipulation that't happening on `generateAndSave` to pass the tid to html meta.

Reply:
https://github.com/wikimedia/restbase/blob/6a8e93390e59552e01d296e0563c9ef3ae1867bb/sys/parsoid.js#L444